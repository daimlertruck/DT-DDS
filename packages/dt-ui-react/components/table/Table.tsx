import { BaseProps } from '../../types';

import {
  TableStyled,
  RowStyled,
  ColumnHeaderStyled,
  CellStyled,
  TableBodyStyled,
  TableHeadStyled,
  TextAlign,
} from './Table.styled';

export interface TableProps extends BaseProps {
  isFixed?: boolean;
}

interface TableRowProps extends BaseProps {
  onClick?: (event: React.MouseEvent<HTMLElement>) => void;
}

interface ColumnHeaderProps extends BaseProps {
  textAlign?: TextAlign;
}

interface DataCellProps extends BaseProps {
  columnWidth?: string;
  textAlign?: TextAlign;
  dataLabel: string;
}

export const Table = ({ children, dataTestId, isFixed }: TableProps) => (
  <TableStyled isFixed={isFixed} data-testid={dataTestId ?? 'table'}>
    {children}
  </TableStyled>
);

export const TableHead = ({ children }: BaseProps) => (
  <TableHeadStyled>{children}</TableHeadStyled>
);

export const TableBody = ({ children }: BaseProps) => {
  return <TableBodyStyled>{children}</TableBodyStyled>;
};

export const TableRow = ({ children, onClick }: TableRowProps) => {
  const isSelectable = Boolean(onClick);

  return (
    <RowStyled {...(onClick && { onClick })} selectableRow={isSelectable}>
      {children}
    </RowStyled>
  );
};

export const ColumnHeader = ({
  children,
  textAlign = 'left',
}: ColumnHeaderProps) => (
  <ColumnHeaderStyled textAlign={textAlign}>{children}</ColumnHeaderStyled>
);

export const DataCell = ({
  children,
  columnWidth,
  dataLabel,
  textAlign = 'left',
}: DataCellProps) => (
  <CellStyled
    data-label={dataLabel}
    columnWidth={columnWidth}
    textAlign={textAlign}
  >
    {children}
  </CellStyled>
);
