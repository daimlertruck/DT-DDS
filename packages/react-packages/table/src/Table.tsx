import { BaseProps } from '@dt-ui/react-core';
import { useEffect, useRef, useState } from 'react';

import {
  TableBody,
  TableColumnHeader,
  TableDataCell,
  TableHead,
  TableRow,
} from './components';
import { TableContextProvider } from './context';
import { TableStyled } from './Table.styled';
import { observeShouldColumnsBeFixed } from './utils/fixedColumns';

export interface TableProps extends BaseProps {
  isFixed?: boolean;
  fixedColumnCount?: number;
  fixedEndColumnCount?: number;
}

const Table = ({
  children,
  dataTestId,
  style,
  isFixed = false,
  fixedColumnCount = 0,
  fixedEndColumnCount = 0,
}: TableProps) => {
  const tableRef = useRef<HTMLTableElement>(null);

  const [isColumnsFixed, setIsColumnsFixed] = useState(false);

  useEffect(() => {
    const observer = observeShouldColumnsBeFixed(
      tableRef,
      fixedColumnCount,
      fixedEndColumnCount,
      setIsColumnsFixed
    );
    return () => observer.disconnect();
  }, [fixedColumnCount, fixedEndColumnCount]);

  return (
    <TableContextProvider
      fixedColumnCount={fixedColumnCount}
      fixedEndColumnCount={fixedEndColumnCount}
      isColumnsFixed={isColumnsFixed}
    >
      <TableStyled
        data-testid={dataTestId ?? 'table'}
        isFixed={isFixed}
        ref={tableRef}
        style={style}
      >
        {children}
      </TableStyled>
    </TableContextProvider>
  );
};

Table.Head = TableHead;
Table.ColumnHeader = TableColumnHeader;
Table.Body = TableBody;
Table.Row = TableRow;
Table.DataCell = TableDataCell;

export default Table;
