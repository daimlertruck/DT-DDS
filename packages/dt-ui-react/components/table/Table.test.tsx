import { Provider, theme } from '@dt-ui/react-core';
import { render, screen, within } from '@testing-library/react';
import '@emotion/jest';

import { default as Table } from './Table';

describe('<Table /> component', () => {
  const COLUMNS = ['Column 1', 'Column 2', 'Column 3', 'Column 4'];
  const ROWS = [
    [
      'Row 1 - content 1',
      'Row 1 - content 2',
      'Row 1 - content 3',
      'Row 1 - content 4',
    ],
    [
      'Row 2 - content 1',
      'Row 2 - content 2',
      'Row 2 - content 3',
      'Row 2 - content 4',
    ],
    [
      'Row 3 - content 1',
      'Row 3 - content 2',
      'Row 3 - content 3',
      'Row 3 - content 4',
    ],
  ];

  it('renders a Composable Table with a custom data structure', () => {
    render(
      <Provider theme={theme}>
        <Table>
          <Table.Head>
            <Table.Row>
              {COLUMNS.map((column: string) => (
                <Table.ColumnHeader key={`column-index-${column}`}>
                  {column}
                </Table.ColumnHeader>
              ))}
            </Table.Row>
          </Table.Head>
          <Table.Body>
            {ROWS.map((row: string[], rowIndex: number) => (
              <Table.Row key={`row-index-${row}`}>
                {row.map((content: string) => (
                  <Table.DataCell
                    dataLabel={COLUMNS[rowIndex]}
                    key={`data-cell-index-${content}`}
                  >
                    {content}
                  </Table.DataCell>
                ))}
              </Table.Row>
            ))}
          </Table.Body>
        </Table>
      </Provider>
    );

    // Test head columns are rendered
    COLUMNS.forEach((column: string) => {
      const cell = screen.getByText(column).closest('th');
      const utils = within(cell!);

      expect(utils.getByText(column)).toBeInTheDocument();
    });

    // Test rows content are in rendered
    ROWS.forEach((row: string[]) => {
      row.forEach((cell: string) => {
        const dataCell = screen.getByText(cell).closest('td');
        const utils = within(dataCell!);

        expect(utils.getByText(cell)).toBeInTheDocument();
      });
    });
  });

  it('renders a Basic Table', () => {
    const { container } = render(
      <Provider theme={theme}>
        <Table>
          <Table.Head>
            <Table.Row>
              {COLUMNS.map((column: string) => (
                <Table.ColumnHeader key={`column-header-${column}`}>
                  {column}
                </Table.ColumnHeader>
              ))}
            </Table.Row>
          </Table.Head>
          <Table.Body>
            {ROWS.map((row: string[]) => (
              <Table.Row key={row.toString()}>
                {row.map((content: string, i: number) => (
                  <Table.DataCell
                    dataLabel={COLUMNS[i]}
                    key={`column-${COLUMNS[i]}-${content}`}
                  >
                    {content}
                  </Table.DataCell>
                ))}
              </Table.Row>
            ))}
          </Table.Body>
        </Table>
      </Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('applies the expected styles when Table has Fixed Header', () => {
    render(
      <Provider theme={theme}>
        <Table.Head hasFixedHeader>
          <Table.Row>
            {COLUMNS.map((column: string) => (
              <Table.ColumnHeader key={`column-header-${column}`}>
                {column}
              </Table.ColumnHeader>
            ))}
          </Table.Row>
        </Table.Head>
        <Table.Body>
          {ROWS.map((row: string[]) => (
            <Table.Row key={row.toString()}>
              {row.map((content: string, i: number) => (
                <Table.DataCell
                  dataLabel={COLUMNS[i]}
                  key={`column-${COLUMNS[i]}-${content}`}
                >
                  {content}
                </Table.DataCell>
              ))}
            </Table.Row>
          ))}
        </Table.Body>
      </Provider>
    );

    const theadElement = screen.getByText('Column 1').closest('thead');

    expect(theadElement).toHaveStyleRule('position', 'sticky', {
      media: `(min-width: ${theme.breakpoints.m})`,
    });
  });
});
