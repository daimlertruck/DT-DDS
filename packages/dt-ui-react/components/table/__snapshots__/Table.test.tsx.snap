// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> component renders a Basic Table 1`] = `
.emotion-0 {
  table-layout: auto;
  width: 100%;
  border-spacing: 0;
}

.emotion-2 {
  display: none;
}

.emotion-2 tr {
  background: transparent;
}

@media (min-width: 900px) {
  .emotion-2 {
    display: table-header-group;
  }
}

.emotion-4 {
  display: block;
  padding: 24px;
  width: 100%;
  background-color: #FFFFFF;
  border-radius: 8px;
  box-shadow: 0px 1px 3px 0px rgba(0, 0, 0, 0.1),0px 1px 2px -1px rgba(0, 0, 0, 0.1);
  cursor: auto;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

@media (min-width: 900px) {
  .emotion-4 {
    display: table-row;
    border-radius: 0px;
    box-shadow: none;
    padding: 0;
  }
}

.emotion-6 {
  text-align: left;
  color: #323232;
  padding: 0 16px 16px;
  font-weight: 600;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
}

.emotion-14 {
  display: grid;
  grid-template-columns: 1fr;
  gap: 12px;
}

.emotion-14 tr:hover {
  background-color: #323232;
}

@media (min-width: 600px) {
  .emotion-14 {
    grid-template-columns: 1fr 1fr;
  }
}

@media (min-width: 900px) {
  .emotion-14 {
    display: table-header-group;
  }
}

.emotion-18 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 12px;
  padding: 12px 16px;
  word-break: break-word;
  text-align: left;
  color: #7C7C7D;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
}

.emotion-18:not(:first-of-type):before {
  content: attr(data-label);
  color: #323232;
  word-break: keep-all;
  font-weight: 600;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
}

.emotion-18:first-of-type {
  color: #323232;
  border-bottom: 1px solid #8B8C8C;
  padding-top: 0;
  font-weight: 700;
  font-size: 0.875rem;
  line-height: 1.25rem;
  letter-spacing: normal;
  margin: 0;
}

@media (min-width: 900px) {
  .emotion-18 {
    display: table-cell;
    border-top: 1px solid #8B8C8C;
    padding: 16px;
    max-width: 0;
    font-weight: 400;
    font-size: 0.875rem;
    line-height: 1.25rem;
    letter-spacing: normal;
    margin: 0;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  .emotion-18:first-of-type {
    color: #7C7C7D;
    padding-top: 16px;
    border-bottom: none;
    font-weight: 400;
    font-size: 0.875rem;
    line-height: 1.25rem;
    letter-spacing: normal;
    margin: 0;
  }

  .emotion-18:before {
    display: none;
  }
}

<div>
  <table
    class="emotion-0 emotion-1"
    data-testid="table"
  >
    <thead
      class="emotion-2 emotion-3"
    >
      <tr
        class="emotion-4 emotion-5"
      >
        <th
          class="emotion-6 emotion-7"
        >
          Column 1
        </th>
        <th
          class="emotion-6 emotion-7"
        >
          Column 2
        </th>
        <th
          class="emotion-6 emotion-7"
        >
          Column 3
        </th>
        <th
          class="emotion-6 emotion-7"
        >
          Column 4
        </th>
      </tr>
    </thead>
    <tbody
      class="emotion-14 emotion-15"
    >
      <tr
        class="emotion-4 emotion-5"
      >
        <td
          class="emotion-18 emotion-19"
          data-label="Column 1"
        >
          Row 1 - content 1
        </td>
        <td
          class="emotion-18 emotion-19"
          data-label="Column 2"
        >
          Row 1 - content 2
        </td>
        <td
          class="emotion-18 emotion-19"
          data-label="Column 3"
        >
          Row 1 - content 3
        </td>
        <td
          class="emotion-18 emotion-19"
          data-label="Column 4"
        >
          Row 1 - content 4
        </td>
      </tr>
      <tr
        class="emotion-4 emotion-5"
      >
        <td
          class="emotion-18 emotion-19"
          data-label="Column 1"
        >
          Row 2 - content 1
        </td>
        <td
          class="emotion-18 emotion-19"
          data-label="Column 2"
        >
          Row 2 - content 2
        </td>
        <td
          class="emotion-18 emotion-19"
          data-label="Column 3"
        >
          Row 2 - content 3
        </td>
        <td
          class="emotion-18 emotion-19"
          data-label="Column 4"
        >
          Row 2 - content 4
        </td>
      </tr>
      <tr
        class="emotion-4 emotion-5"
      >
        <td
          class="emotion-18 emotion-19"
          data-label="Column 1"
        >
          Row 3 - content 1
        </td>
        <td
          class="emotion-18 emotion-19"
          data-label="Column 2"
        >
          Row 3 - content 2
        </td>
        <td
          class="emotion-18 emotion-19"
          data-label="Column 3"
        >
          Row 3 - content 3
        </td>
        <td
          class="emotion-18 emotion-19"
          data-label="Column 4"
        >
          Row 3 - content 4
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
